server/store/session.go

```diff
@@ -0,0 +1,170 @@
+package store
+
+import (
+	"strings"
+
+	pluginapi "github.com/mattermost/mattermost-plugin-api"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/server/config"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+)
+
+type SessionStore interface {
+	Get(appID apps.AppID, userID string) (*model.Session, error)
+	ListForApp(appID apps.AppID) ([]*model.Session, error)
+	ListForUser(userID string) ([]*model.Session, error)
+	Save(appID apps.AppID, userID string, session *model.Session) error
+	Delete(appID apps.AppID, userID string) error
+	DeleteAllForApp(appID apps.AppID) error
+}
+
+type sessionStore struct {
+	*Service
+}
+
+var _ SessionStore = (*sessionStore)(nil)
+
+func sessionKey(appID apps.AppID, userID string) string {
+	return appKey(appID) + "_" + userID
+}
+
+func appKey(appID apps.AppID) string {
+	return config.KVTokenPrefix + "_" + string(appID)
+}
+
+func parseKey(key string) (apps.AppID, string, error) {
+	s := strings.Split(key, "_")
+	if len(s) != 3 {
+		return "", "", errors.New("invalid key pattern")
+	}
+
+	if s[0] != config.KVTokenPrefix {
+		return "", "", errors.New("invalid key prefix")
+	}
+
+	return apps.AppID(s[1]), s[2], nil
+}
+
+func (s sessionStore) Get(appID apps.AppID, userID string) (*model.Session, error) {
+	return s.get(sessionKey(appID, userID))
+}
+
+func (s sessionStore) get(key string) (*model.Session, error) {
+	var session model.Session
+	err := s.conf.MattermostAPI().KV.Get(key, &session)
+	if err != nil {
+		return nil, err
+	}
+
+	if session.Id == "" {
+		return nil, utils.ErrNotFound
+	}
+
+	return &session, nil
+}
+
+func (s sessionStore) Save(appID apps.AppID, userID string, session *model.Session) error {
+	_, err := s.conf.MattermostAPI().KV.Set(sessionKey(appID, userID), session)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s sessionStore) listKeysForApp(appID apps.AppID) ([]string, error) {
+	ret := make([]string, 0)
+
+	for i := 0; ; i++ {
+		keys, err := s.conf.MattermostAPI().KV.ListKeys(i, keysPerPage, pluginapi.WithPrefix(appKey(appID)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list keys - page, %d", i)
+		}
+
+		ret = append(ret, keys...)
+
+		if len(keys) < keysPerPage {
```

Comment:
I've heard that the database does not behave that beautifully in all cases. I think the potential problem was with other search engines like Elastic Search, where it may return less results than the perPage value, and still have more results to give. Since this part of the database is not using the search engines, probably it is fine, but it would be nice to have a second opinion.

@mgdelacroix Do you know if we may have that problem here?

Reply:
FWIW: This patters is used in multiple plugins like in https://github.com/mattermost/mattermost-plugin-api/blob/cd302a7195a0ed31ca8d6358ef2bb4eb32a6926d/cluster/job_once_scheduler.go#L118-L120

Reply:
The problem that you're describing happens when the search engine and the database differ. You may get 20 results from ES, one of which should have been deleted from the index and will not be returned when we go to the database to get the information on those 20 results, effectively yielding 19.

There are more results and more pages to go through, but we're returning an amount inferior to the page, so the pattern we changed this with is to keep looping on pages until you get zero results. In this case, if we're only using the database, we don't need to cover for the case, but that's the whole explanation for this.